File: 2121-FindIfPathExistsInGraph/2121-FindIfPathExistsInGraph.java
// Last updated: 8/13/2025, 8:16:39 PM
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        ArrayList<ArrayList<Integer>> graph = new ArrayList();
        for(int i=0;i<n;i++){
            graph.add(new ArrayList());
        }
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            graph.get(u).add(v);
            graph.get(v).add(u);
        }
        boolean[] visited = new boolean[n];
        // visited[source] = true;
        validPathHelper(graph,visited,source);
        return visited[destination];
    }

    public void validPathHelper(ArrayList<ArrayList<Integer>> graph, boolean[] visited, int src){
        if (visited[src]) return;
        visited[src] = true;
        for (int child : graph.get(src)) {
            if (!visited[child])
                validPathHelper(graph,visited,child);
        }
    }
}

File: 2121-FindIfPathExistsInGraph/2121-FindIfPathExistsInGraph.py
# Last updated: 8/13/2025, 8:16:34 PM
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        adj = self.convertToAdjMatrix(n , edges)
        
        visited = [False]*(len(adj))
        my_queue = []
        my_queue.append(adj[0])
        visited[0] = True

        while len(my_queue) > 0 : 
            neighbours = my_queue.pop(0)

            for n in neighbours : 
                if visited[n] != True : 
                    my_queue.append(adj[n])
                    visited[n] = True
        
        return visited[source] and visited[destination]
    
    def convertToAdjMatrix(self , n : int , edges : List[List[int]]) -> List[List[int]]: 
        adj = []

        for i in range (n) : 
            adj.append([]) #intializing list of list
        
        for u,v in edges : 
            adj[u].append(v)
            adj[v].append(u)
        
        return adj